File: cpp/common/normal_distribution.cpp
#include "normal_distribution.h"
#include <cmath>

namespace quant {

double standard_normal_cdf(double x) {
    // C++17 onwards provides std::erfc.
    // Phi(x) = 0.5 * erfc(-x / sqrt(2))
    return 0.5 * std::erfc(-x / std::sqrt(2.0));
}

double standard_normal_pdf(double x) {
    static const double inv_sqrt_2pi = 1.0 / std::sqrt(2.0 * M_PI);
    return inv_sqrt_2pi * std::exp(-0.5 * x * x);
}

}  // namespace quant

File: python/common/normal_distribution.py
"""
Standard normal distribution utilities.

This module provides convenience wrappers around the standard normal
distribution functions used throughout the portfolio (e.g. in the
Black-Scholes pricing formula).
"""

import math


def standard_normal_cdf(x: float) -> float:
    """
    Cumulative distribution function of the standard normal N(0, 1).

    Parameters
    ----------
    x : float
        Point at which to evaluate the CDF.

    Returns
    -------
    float
        Phi(x) = P(Z <= x) where Z ~ N(0, 1).
    """
    return 0.5 * (1.0 + math.erf(x / math.sqrt(2.0)))


def standard_normal_pdf(x: float) -> float:
    """
    Probability density function of the standard normal N(0, 1).
    """
    return math.exp(-0.5 * x * x) / math.sqrt(2.0 * math.pi)


if __name__ == "__main__":
    # Quick sanity check
    test_points = [-3.0, -1.0, 0.0, 1.0, 3.0]
    print("Standard normal CDF values:")
    for x in test_points:
        print(f"  Phi({x:+.1f}) = {standard_normal_cdf(x):.6f}")